nixos/modules/services/networking/rosenpass.nix

```diff
@@ -0,0 +1,212 @@
+            }
+          ];
+          relevantExtractions = filter (x: x.relevant) extractions;
+          extract = { root, peer, key, ... }:
+            filter (x: x != null) (flatten (concatMap (x: (map key (peer x))) (attrValues root)));
+          configuredKeys = flatten (map extract relevantExtractions);
+          itemize = xs: concatLines (map (x: " - ${x}") xs);
+          descriptions = map (x: "`${x.description}`");
+          missingKeys = filter (key: !builtins.elem key configuredKeys) (map (x: x.peer) cfg.settings.peers);
+          unusual = ''
+            While this may work as expected, e.g. you want to manually configure WireGuard,
+            such a scenario is unusual. Please double-check your configuration.
+          '';
+        in
+        (optional (relevantExtractions != [ ] && missingKeys != [ ]) ''
+          You have configured Rosenpass peers with the WireGuard public keys:
+          ${itemize missingKeys}
+          But there is no corresponding active Wireguard peer configuration in any of:
+          ${itemize (descriptions relevantExtractions)}
+          ${unusual}
+        '')
+        ++
+        (optional (relevantExtractions == [ ]) ''
+          You have configured Rosenpass, but you have not configured Wireguard via any of:
+          ${itemize (descriptions extractions)}
+          ${unusual}
+        '');
+
+      environment.systemPackages = [ cfg.package pkgs.wireguard-tools ];
+
+      systemd.services.rosenpass =
+        let
+          filterNonNull = filterAttrsRecursive (_: v: v != null);
+          config = settingsFormat.generate "config.toml" (
+            filterNonNull (cfg.settings
+              //
+              (
+                let
+                  credentialPath = id: "$CREDENTIALS_DIRECTORY/${id}";
+                  # NOTE: We would like to remove all `null` values inside `cfg.settings`
+                  # recursively, since `settingsFormat.generate` cannot handle `null`.
+                  # This would require to traverse both attribute sets and lists recursively.
+                  # `filterAttrsRecursive` only recurses into attribute sets, but not
+                  # into values that might contain other attribute sets (such as lists,
+                  # e.g. `cfg.settings.peers`). Here, we just specialize on `cfg.settings.peers`,
+                  # and this may break unexpectedly whenever a `null` value is contained
+                  # in a list in `cfg.settings`, other than `cfg.settings.peers`.
+                  peersWithoutNulls = map filterNonNull cfg.settings.peers;
+                  # NOTE: This works around a bug/feature in Rosenpass.
+                  # It will ignore peers that do not have `extra_params`, and an empty
+                  # list will prevent ignoring.
+                  peersWithExtraParams = map (peer: peer // { extra_params = peer.extra_params or []; }) peersWithoutNulls;
```

Comment:
Since `peer` is a freeform submodule, I think you could define an `extra_params` option with a default value for the same effect. (It's okay as-is too.)

Reply:
I removed this just yesterday. Please see: https://github.com/NixOS/nixpkgs/pull/254813#discussion_r1337548680

Reply:
Ahh, right, it's that one. This way is fine as-is I think!